autoformer time series blog post (#1184)
huggingface/blog: autoformer.md

In time series analysis, [decomposition](https://en.wikipedia.org/wiki/Decomposition_of_time_series) is a method of breaking down a time series into three systematic components, including trend-cycle, seasonal variation, and random fluctuations.
The trend component represents the long-term direction of the time series, which can be increasing, decreasing, or stable over time. The seasonal component represents the recurring patterns that occur within the time series, such as yearly or quarterly cycles. Finally, the random (sometimes called "irregular") component represents the random noise in the data that cannot be explained by the trend or seasonal components. 

Two main types of decomposition are additive and multiplicative decomposition, which are implemented in the [great statsmodels library](https://www.statsmodels.org/dev/generated/statsmodels.tsa.seasonal.seasonal_decompose.html). By decomposing a time series into these components, we can better understand and model the underlying patterns in the data. 

But how can we incorporate decomposition into the Transformer architecture? Let's see how Autoformer does it.

#### Decomposition in Autoformer

> | ![autoformer_architecture](https://huggingface.co/datasets/huggingface/documentation-images/resolve/main/blog/148_autoformer/autoformer_architecture.png) |

Reviewer: I ran the images through a PNG optimizer, hope you don't mind :) Feel free to merge them in the datasets repo whenever you are done with them.

|:--:|
| Autoformer architecture from [the paper](https://arxiv.org/abs/2106.13008) |

> Autoformer incorporates a decomposition block as an inner operation of the model, as presented in the Autoformer's architecture above. As can be seen, the encoder and decoder use a decomposition block to aggregate the trend-cyclical part and extract the seasonal part from series progressively. The concept of inner decomposition has demonstrated its usefulness since the publication of Autoformer. Subsequently, it has been adopted in several other time series papers, such as FEDformer ([Zhou, Tian, et al., ICML 2022](https://arxiv.org/abs/2201.12740)) and DLinear [(Zeng, Ailing, et al., AAAI 2023)](https://arxiv.org/abs/2205.13504), highlighting its significance in time series modeling.

Reviewer: ```suggestion
Autoformer incorporates a decomposition block as an inner operation of the model, as presented in the Autoformer's architecture above. As can be seen, the encoder and decoder use a decomposition block to aggregate the trend-cyclical part and extract the seasonal part from the series progressively. The concept of inner decomposition has demonstrated its usefulness since the publication of Autoformer. Subsequently, it has been adopted in several other time series papers, such as FEDformer ([Zhou, Tian, et al., ICML 2022](https://arxiv.org/abs/2201.12740)) and DLinear [(Zeng, Ailing, et al., AAAI 2023)](https://arxiv.org/abs/2205.13504), highlighting its significance in time series modeling.
```

maybe?

In the following sections, we will dive into these topics in detail and explain them with code examples.

####  Frequency Domain Attention 

| ![autoformer_autocorrelation_only_attention](https://huggingface.co/datasets/huggingface/documentation-images/resolve/main/blog/148_autoformer/autoformer_autocorrelation_only_attention.png) |
|:--:|
| Attention weights computation in frequency domain using FFT, from [the paper](https://arxiv.org/abs/2106.13008) |

In theory, given a time lag \\(\tau\\), _autocorrelation_ for a single discrete variable \\(y\\) is used to measure the "relationship" (pearson correlation) between the variable's current value at time \\(t\\) to its past value at time \\(t-\tau\\):

$$
\textrm{Autocorrelation}(\tau) = \textrm{Corr}(y_t, y_{t-\tau})
$$

Using autocorrelation, Autoformer extracts frequency-based dependencies from the queries and keys, instead of the standard dot-product between them. You can think about it as a replacement for the \\(QK^T\\) term in the self-attention.

> In practice, autocorrelation of the queries and keys for **all lags** is calculated at once by FFT. By doing so, the autocorrelation mechanism achives \\(O(L \log L)\\) time complexity (\\(L\\) is the input time length), similar to [Informer's ProbSparse attention](https://huggingface.co/blog/informer#probsparse-attention). Note that the theory behind computing autocorrelation using FFT is based on the [Wiener–Khinchin theorem](https://en.wikipedia.org/wiki/Wiener%E2%80%93Khinchin_theorem), which is outside the scope of this blog post.

Reviewer: ```suggestion
In practice, autocorrelation of the queries and keys for **all lags** is calculated at once by FFT. By doing so, the autocorrelation mechanism achieves \\(O(L \log L)\\) time complexity (\\(L\\) is the input time length), similar to [Informer's ProbSparse attention](https://huggingface.co/blog/informer#probsparse-attention). Note that the theory behind computing autocorrelation using FFT is based on the [Wiener–Khinchin theorem](https://en.wikipedia.org/wiki/Wiener%E2%80%93Khinchin_theorem), which is outside the scope of this blog post.
```

Now, we are ready to see the code in PyTorch: 

```python
import torch 

def autocorrelation(query_states, key_states):
    """
    Computes autocorrelation(Q,K) using `torch.fft`. 
    Think about it as a replacement for the QK^T in the self-attention.
    
    Assumption: states are resized to same shape of [batch_size, time_length, embedding_dim].
    """
    query_states_fft = torch.fft.rfft(query_states, dim=1)
    key_states_fft = torch.fft.rfft(key_states, dim=1)
    attn_weights = query_states_fft * torch.conj(key_states_fft)
    attn_weights = torch.fft.irfft(attn_weights, dim=1)  
    
    return attn_weights
```

Quite simple! 😎 Please be aware that this is only a partial implementation of `autocorrelation(Q,K)`, and the full implementation can be found in 🤗 Transformers.

> Next, we will see how to aggregate our `attn_weights` with the values by time delay, which are termed as _Time Delay Aggregation_.

Reviewer 2: ```suggestion
Next, we will see how to aggregate our `attn_weights` with the values by time delay, process which is termed as _Time Delay Aggregation_.
```

#### Time Delay Aggregation
| ![autoformer_autocorrelation_only_aggregation](https://huggingface.co/datasets/huggingface/documentation-images/resolve/main/blog/148_autoformer/autoformer_autocorrelation_only_aggregation.png) |
|:--:|

> | Aggregation by time delay, from [the paper](https://arxiv.org/abs/2106.13008) |

Reviewer: ```suggestion
| Aggregation by time delay, from [the Autoformer paper](https://arxiv.org/abs/2106.13008) |
```

(ignore, of course, if you think it's too redundant)

    """
    bsz, num_heads, tgt_len, channel = ...
    time_length = value_states.size(1)
    autocorrelations = attn_weights.view(bsz, num_heads, tgt_len, channel)

    # find top k autocorrelations delays
    top_k = int(autocorrelation_factor * math.log(time_length))
    autocorrelations_mean = torch.mean(autocorrelations, dim=(1, -1))  # bsz x tgt_len
    top_k_autocorrelations, top_k_delays = torch.topk(autocorrelations_mean, top_k, dim=1)

    # apply softmax on the channel dim
    top_k_autocorrelations = torch.softmax(top_k_autocorrelations, dim=-1)  # bsz x top_k

    # compute aggregation: value_states.roll(delay) * top_k_autocorrelations(delay)
    delays_agg = torch.zeros_like(value_states).float()  # bsz x time_length x channel
    for i in range(top_k):
        value_states_roll_delay = value_states.roll(shifts=-int(top_k_delays[i]), dims=1)
        top_k_at_delay = top_k_autocorrelations[:, i]
        # aggregation
        top_k_resized = top_k_at_delay.view(-1, 1, 1).repeat(num_heads, tgt_len, channel)
        delays_agg += value_states_roll_delay * top_k_resized

    attn_output = delays_agg.contiguous()
    return attn_output
```
   

> We did it! The Autoformer model in [now available](https://huggingface.co/docs/transformers/main/en/model_doc/autoformer) in the 🤗 Transformers library, and simply called `AutoformerModel`.

Reviewer: ```suggestion
We did it! The Autoformer model is [now available](https://huggingface.co/docs/transformers/main/en/model_doc/autoformer) in the 🤗 Transformers library, and simply called `AutoformerModel`.
```

Our strategy with this model is to show the performance of the univariate Transformer models in comparison to the DLinear model which is inherently univariate. We will also present the results from _two_ multivariate Transformer models trained on the same data.

## DLinear 

> The DLinear model uses the above `DecompositionLayer` to  decomposes the input time series into the residual (the seasonality) and trend part.  In the forward then each part is passed through its own linear layer which projects the signal to an appropriate `prediction_length` sized output. The final output is the sum of the two corresponding outputs in the point-forecasting model:

Reviewer: ```suggestion
The DLinear model uses the above `DecompositionLayer` to  decompose the input time series into the residual (the seasonality) and trend part. In the forward pass each part is passed through its own linear layer, which projects the signal to an appropriate `prediction_length`-sized output. The final output is the sum of the two corresponding outputs in the point-forecasting model:
```

```python
def forward(self, context):

>     res, trend = self.decomposition(context)
>     seasonal_output = self.linear_seasonal(res)

Reviewer 2: ```suggestion
    seasonal, trend = self.decomposition(context)
    seasonal_output = self.linear_seasonal(seasonal)
```

    trend_output = self.linear_trend(trend)
    return seasonal_output + trend_output
```

In the probabilistic setting one can project the context length arrays to  `prediction-length * hidden` dimensions via the `linear_seasonal` and `linear_trend` layers.  The resulting outputs are added and reshaped to `(prediction_length, hidden)`. Finally, a probabilistic head maps the latent representations of size `hidden` to the parameters of some distribution.

> For our benchmark, we will use the implementation of DLinear from GluonTS](https://github.com/awslabs/gluonts).

Reviewer 2: ```suggestion
For our benchmark, we will use the implementation of DLinear from [GluonTS](https://github.com/awslabs/gluonts).
```

## Benchmarking - Transformers vs. DLinear

We want to show empirically the performance of Transformer based models, by benchmarking on the `traffic` dataset, in the univariate setting. We will keep the following hyper-parameters fixed for all the models:
* `context_length = prediction_length*2`
* `batch_size = 128`
* `num_batches_per_epoch = 100`
* `epochs = 50`
* `scaling = "std"`

The transformers models are all relatively small  with:
* `encoder_layers=2`
* `decoder_layers=2`
* `d_model=16`

> Instead of showing how to train a model using `Autoformer` one can just replace the model in the previous two blog posts with the new `Autoformer` model and train it on the `traffic` dataset. In order to not repeat ourself, we  have done train the models already and pushed them to the Hugging Face Hub. In the following we will use the already saved models for just the evaluations.

Reviewer: ```suggestion
Instead of showing how to train a model using `Autoformer` one can just replace the model in the previous two blog posts with the new `Autoformer` model and train it on the `traffic` dataset. In order to not repeat ourselves, we  have already trained the models and pushed them to the Hugging Face Hub. We will use those models for evaluation.
```

(not sure what the "previous two blog posts" refer to, is the reader required to be familiar with them?)

Reviewer 2: Good point, thanks! will add the [Informer](https://huggingface.co/blog/informer) and [TimeSeriesTransformer](https://huggingface.co/blog/time-series-transformers) blogs here

        periodicity=get_seasonality(freq))
    mase_metrics.append(mase["mase"])
```

```python
print(f"MASE: {np.mean(mase_metrics)}")

>>> MASE: 0.9092493026949344
```

To plot the prediction for any time series with respect to the ground truth test data we define the following helper:

```python
import matplotlib.dates as mdates
import pandas as pd

test_ds = list(test_dataset)

def plot(ts_index):
    fig, ax = plt.subplots()

    index = pd.period_range(
        start=test_ds[ts_index][FieldName.START],
        periods=len(test_ds[ts_index][FieldName.TARGET]),
        freq=test_ds[ts_index][FieldName.START].freq,
    ).to_timestamp()

    ax.plot(
        index[-5*prediction_length:], 
        test_ds[ts_index]["target"][-5*prediction_length:],
        label="actual",
    )

    plt.plot(
        index[-prediction_length:], 
        np.median(forecasts[ts_index], axis=0),
        label="median",
    )
    
    plt.gcf().autofmt_xdate()
    plt.legend(loc="best")
    plt.show()
```

For example:

```python
plot(4)
```
    
![png](https://huggingface.co/datasets/huggingface/documentation-images/resolve/main/blog/148_autoformer/output_44_0.png)
    

## DLinear

> A probablistic DLinear is implemented in `glounts` and thus we can train and evaluate it relatively quickly here:

Reviewer: ```suggestion
A probablistic DLinear is implemented in `gluonts` and thus we can train and evaluate it relatively quickly here:
```

```python
from gluonts.torch.model.d_linear.estimator import DLinearEstimator

estimator = DLinearEstimator(
    prediction_length=dataset.metadata.prediction_length,
    context_length=dataset.metadata.prediction_length*2,
    scaling="std",
    hidden_dimension=2,
    
    batch_size=128,
    num_batches_per_epoch=100,
    trainer_kwargs=dict(max_epochs=50)
)
```

```python
predictor = estimator.train(
    training_data=train_dataset, 
    cache_data=True, 
    shuffle_buffer_length=1024
)

> >>> Authorization required, but no authorization protocol specified

Reviewer 2: @kashif maybe we can remove this output?

            # sort of running counter
            AddAgeFeature(
                target_field=FieldName.TARGET,
                output_field=FieldName.FEAT_AGE,
                pred_length=config.prediction_length,
                log_scale=True,
            ),
            # step 6: vertically stack all the temporal features into the key FEAT_TIME
            VstackFeatures(
                output_field=FieldName.FEAT_TIME,
                input_fields=[FieldName.FEAT_TIME, FieldName.FEAT_AGE]
                + (
                    [FieldName.FEAT_DYNAMIC_REAL]
                    if config.num_dynamic_real_features > 0
                    else []
                ),
            ),
            # step 7: rename to match HuggingFace names
            RenameFields(
                mapping={
                    FieldName.FEAT_STATIC_CAT: "static_categorical_features",
                    FieldName.FEAT_STATIC_REAL: "static_real_features",
                    FieldName.FEAT_TIME: "time_features",
                    FieldName.TARGET: "values",
                    FieldName.OBSERVED_VALUES: "observed_mask",
                }
            ),
        ]
    )
```

> ## Define `InstanceSplitter`

Reviewer 2: done in https://github.com/huggingface/blog/pull/1184/commits/640716b2e1fb6d5346b7787cc986c64f3da28f9a

    instance_sampler = create_instance_splitter(config, "test")

    # we apply the transformations in test mode
    testing_instances = instance_sampler.apply(transformed_data, is_train=False)

    return as_stacked_batches(
        testing_instances,
        batch_size=batch_size,
        output_type=torch.tensor,
        field_names=PREDICTION_INPUT_NAMES,
    )
```

## Evaluate on Autoformer

We have already pre-trained an Autoformer model on this dataset so we can just fetch the model and evaluate it on the test set:

```python
from transformers import AutoformerConfig, AutoformerForPrediction
```

```python
config = AutoformerConfig.from_pretrained("kashif/autoformer-traffic-hourly")
```

```python
test_dataloader = create_test_dataloader(
    config=config,
    freq=freq,
    data=test_dataset,
    batch_size=64,
)
```

At inference time, we will use the model's  `generate()` method for predicting `predection_length` into the future from the very last context window of each time series in the training set.  

```python
model = AutoformerForPrediction.from_pretrained("kashif/autoformer-traffic-hourly")
```

```python

> from accelerate import Accelerator

Reviewer 2: Why we need here the:

```python
accelerator = Accelerator()
device = accelerator.device
model.to(device)

model.eval()

```

In the Informer we just started with 

```python
model.eval()

forecasts_ = []
```

accelerator = Accelerator()
device = accelerator.device
model.to(device)

model.eval()

forecasts_ = []

for batch in test_dataloader:
    outputs = model.generate(
        static_categorical_features=batch["static_categorical_features"].to(device)
        if config.num_static_categorical_features > 0
        else None,
        static_real_features=batch["static_real_features"].to(device)
        if config.num_static_real_features > 0
        else None,
        past_time_features=batch["past_time_features"].to(device),
        past_values=batch["past_values"].to(device),
        future_time_features=batch["future_time_features"].to(device),
        past_observed_mask=batch["past_observed_mask"].to(device),
    )
    forecasts_.append(outputs.sequences.cpu().numpy())
```

> The model outputs a tensor of shape (`batch_size`, `number of samples`, `prediction length`, `input_size`). 

Reviewer 2: I think this line is irrelevant @kashif

```python
import numpy as np

forecasts = np.vstack(forecasts_)
print(forecasts.shape)

>>> (6034, 100, 24)
```

We can evaluate the resulting forecast with respect to the ground truth out of sample values present in the test set. For that, we'll use the 🤗 [Evaluate](https://huggingface.co/docs/evaluate/index) library, which includes the [MASE](https://huggingface.co/spaces/evaluate-metric/mase) metrics.

We calculate the metric for each time series in the dataset and return the average:

```python
from tqdm.autonotebook import tqdm
from evaluate import load
from gluonts.time_feature import get_seasonality

mase_metric = load("evaluate-metric/mase")

forecast_median = np.median(forecasts, 1)

mase_metrics = []
for item_id, ts in enumerate(tqdm(test_dataset)):
    training_data = ts["target"][:-prediction_length]
    ground_truth = ts["target"][-prediction_length:]
    mase = mase_metric.compute(
        predictions=forecast_median[item_id], 
        references=np.array(ground_truth), 
        training=np.array(training_data), 
        periodicity=get_seasonality(freq))
    mase_metrics.append(mase["mase"])
```

```python

> print(f"MASE: {np.mean(mase_metrics)}")

Reviewer 2: Add print to 3 digits

Reviewer 2: Done in https://github.com/huggingface/blog/pull/1184/commits/2e8c465c50eda379dc4dd49087e3a22d453906ab

    index = pd.period_range(
        start=test_ds[ts_index][FieldName.START],
        periods=len(test_ds[ts_index][FieldName.TARGET]),
        freq=test_ds[ts_index][FieldName.START].freq,
    ).to_timestamp()

    ax.plot(
        index[-5*prediction_length:], 
        test_ds[ts_index]["target"][-5*prediction_length:],
        label="actual",
    )

    plt.plot(
        index[-prediction_length:], 
        np.median(forecasts[ts_index], axis=0),
        label="median",
    )
    
    plt.gcf().autofmt_xdate()
    plt.legend(loc="best")
    plt.show()
```

For example:

```python
plot(4)
```

![png](https://huggingface.co/datasets/huggingface/documentation-images/resolve/main/blog/148_autoformer/output_44_0.png)
    

## Evaluate on DLinear

A probablistic DLinear is implemented in `gluonts` and thus we can train and evaluate it relatively quickly here:

```python
from gluonts.torch.model.d_linear.estimator import DLinearEstimator

estimator = DLinearEstimator(
    prediction_length=dataset.metadata.prediction_length,
    context_length=dataset.metadata.prediction_length*2,
    scaling="std",
    hidden_dimension=2,
    
    batch_size=128,
    num_batches_per_epoch=100,

>     trainer_kwargs=dict(max_epochs=50)

Reviewer 2: Here, numbers can be changed to the corresponding variables

Reviewer 2: done in https://github.com/huggingface/blog/pull/1184/commits/62373c3d6c1cd6c72222d51702cf02efd144b3dc

      rank_zero_warn(
    INFO:pytorch_lightning.callbacks.model_summary:
      | Name  | Type         | Params
    ---------------------------------------
    0 | model | DLinearModel | 4.7 K 
    ---------------------------------------
    4.7 K     Trainable params
    0         Non-trainable params
    4.7 K     Total params
    0.019     Total estimated model params size (MB)

    Training: 0it [00:00, ?it/s]
    ...
    INFO:pytorch_lightning.utilities.rank_zero:Epoch 49, global step 5000: 'train_loss' was not in top 1
    INFO:pytorch_lightning.utilities.rank_zero:`Trainer.fit` stopped: `max_epochs=50` reached.
```

```python
from gluonts.evaluation import make_evaluation_predictions, Evaluator

forecast_it, ts_it = make_evaluation_predictions(
    dataset=dataset.test,
    predictor=predictor,
)
```

```python
d_linear_forecasts = list(forecast_it)
d_linear_tss = list(ts_it)
```

```python
evaluator = Evaluator()

agg_metrics, _ = evaluator(iter(d_linear_tss), iter(d_linear_forecasts))
```

```python

> print(agg_metrics["MASE"])

Reviewer 2: same, print to 3 digits

Reviewer 2: Done in https://github.com/huggingface/blog/pull/1184/commits/2e8c465c50eda379dc4dd49087e3a22d453906ab

---
title: "Yes, Transformers are Effective for Time Series Forecasting" 
thumbnail: /blog/assets/148_autoformer/thumbnail.png
authors:
- user: kashif
- user: elisim
  guest: true
- user: nielsr
---

# Yes, Transformers are Effective for Time Series Forecasting

<!-- {blog_metadata} -->
<!-- {authors} -->

## Introduction

A few months ago, we introduced the [Informer](https://huggingface.co/blog/informer) model ([Zhou, Haoyi, et al., 2021](https://arxiv.org/abs/2012.07436)), which is a Time Series Transformer that won the AAAI 2021 best paper award. We also provided an example for multivariate probabilistic forecasting with Informer. In this post, we introduce the _Autoformer_ model ([Wu, Haixu, et al., 2021](https://arxiv.org/abs/2106.13008)), which was published in NeurIPS 2021 after the Informer model. The Autoformer model is [now available](https://huggingface.co/docs/transformers/main/en/model_doc/autoformer) in 🤗 Transformers.

> We will begin by explaining the Autoformer model, including its novel _Decomposition Layer_ and _Attention (Autocorrelation) Mechanism_. Next, we will present the _DLinear_ model, which is a simple feedforward network that uses the decomposition layer from Autoformer. The DLinear model was first introduced in [Are Transformers Effective for Time Series Forecasting?](https://arxiv.org/abs/2205.13504) which published recently in AAAI 2023. Finally, we will provide empirical evidence that **Transformers are indeed Effective for Time Series Forecasting**. Our comparison shows that a simple linear model is not better than Transformers, as claimed. When compared against equivalent sized models in the same setting as the the linear models, the Transformer based models perform better on the test set metrics we consider.

Reviewer 3: ```suggestion
We will begin by explaining the Autoformer model, including its novel _Decomposition Layer_ and _Attention (Autocorrelation) Mechanism_. Next, we will present the _DLinear_ model, which is a simple feedforward network that uses the decomposition layer from Autoformer. The DLinear model was first introduced in [Are Transformers Effective for Time Series Forecasting?](https://arxiv.org/abs/2205.13504) which published recently in AAAI 2023. Finally, we will provide empirical evidence that **Transformers are indeed Effective for Time Series Forecasting**. Our comparison shows that a simple linear model is not better than Transformers, as claimed. When compared against equivalent sized models in the same setting as the linear models, the Transformer based models perform better on the test set metrics we consider.
```

## Autoformer - Under The Hood

The Autoformer builds upon the traditional method of decomposing time series into seasonality and trend-cycle components. This is achieved through the incorporation of a _Decomposition Layer_, which enhances the model's ability to capture these components accurately. Moreover, the Autoformer introduces an innovative auto-correlation mechanism that replaces the standard self-attention used in the vanilla transformer. This mechanism enables the model to utilize period-based dependencies in the attention, thus improving the overall performance. 

> In the upcoming sections, we will delve into the two key contributions of the Autoformer: the _Decomposition Layer_ and the _Attention (Autocorrelation) Mechanism_. We will also provide code examples to illustrate how these components function within the Autoformer architecture.

Reviewer 3: ```suggestion
In the upcoming sections, we will delve into the two key contributions of Autoformer: the _Decomposition Layer_ and the _Attention (Autocorrelation) Mechanism_. We will also provide code examples to illustrate how these components function within the Autoformer architecture.
```
Let's just say "Autoformer" everywhere :)

### Decomposition Layer
Decomposition has long been a popular method in time series analysis, but it had not been extensively incorporated into deep learning models until the introduction of the Autoformer paper. In the upcoming sections, we will introduce the general decomposition concept in time series analysis. Following that, we will demonstrate how this idea is applied in Autoformer using PyTorch code.

####  Decomposition of Time Series
In time series analysis, [decomposition](https://en.wikipedia.org/wiki/Decomposition_of_time_series) is a method of breaking down a time series into three systematic components, including trend-cycle, seasonal variation, and random fluctuations.
The trend component represents the long-term direction of the time series, which can be increasing, decreasing, or stable over time. The seasonal component represents the recurring patterns that occur within the time series, such as yearly or quarterly cycles. Finally, the random (sometimes called "irregular") component represents the random noise in the data that cannot be explained by the trend or seasonal components. 

> Two main types of decomposition are additive and multiplicative decomposition, which implemented in the [great statsmodels library](https://www.statsmodels.org/dev/generated/statsmodels.tsa.seasonal.seasonal_decompose.html). By decomposing a time series into these components, we can better understand and model the underlying patterns in the data. 

Reviewer 3: ```suggestion
Two main types of decomposition are additive and multiplicative decomposition, which are implemented in the [great statsmodels library](https://www.statsmodels.org/dev/generated/statsmodels.tsa.seasonal.seasonal_decompose.html). By decomposing a time series into these components, we can better understand and model the underlying patterns in the data. 
```

    def forward(self, x):
        """Input shape: Batch x Time x EMBED_DIM"""
        # padding on the both ends of time series
        num_of_pads = (self.kernel_size - 1) // 2
        front = x[:, 0:1, :].repeat(1, num_of_pads, 1)
        end = x[:, -1:, :].repeat(1, num_of_pads, 1)
        x_padded = torch.cat([front, x, end], dim=1)

        # calculate the trend and seasonal part of the series
        x_trend = self.avg(x_padded.permute(0, 2, 1)).permute(0, 2, 1)
        x_seasonal = x - x_trend
        return x_seasonal, x_trend
```

As you can see, the implementation is quite simple and can be used in other models, as we will see with DLinear. Now, let's explain the second contribution - _Attention (Autocorrelation) Mechanism_.

### Attention (Autocorrelation) Mechanism

| ![autoformer_autocorrelation_vs_full_attention](https://huggingface.co/datasets/huggingface/documentation-images/resolve/main/blog/148_autoformer/autoformer_autocorrelation_vs_full_attention.png) |
|:--:|
|  Vanilla self attention vs Autocorrelation mechanism, from [the paper](https://arxiv.org/abs/2106.13008) |

> In addition to the decomposition layer, Autoformer employs a novel auto-correlation mechanism which replaces the self-attention seamlessly. In the [vanilla time series transformer](https://huggingface.co/docs/transformers/model_doc/time_series_transformer), attention weights are computed in the time domain and point-wise aggregated. On the other hand, as can be seen in the figure above, in the auto-correlation mechanism they are computed in the frequency domain (using [fast fourier transform](https://en.wikipedia.org/wiki/Fast_Fourier_transform)) and aggregated by time delay. 

Reviewer 3: ```suggestion
In addition to the decomposition layer, Autoformer employs a novel auto-correlation mechanism which replaces the self-attention seamlessly. In the [vanilla Time Series Transformer](https://huggingface.co/docs/transformers/model_doc/time_series_transformer), attention weights are computed in the time domain and point-wise aggregated. On the other hand, as can be seen in the figure above, they are computed in the frequency domain (using [fast fourier transform](https://en.wikipedia.org/wiki/Fast_Fourier_transform)) and aggregated by time delay in case of the auto-correlation mechanism.
```

    # find top k autocorrelations delays
    top_k = int(autocorrelation_factor * math.log(time_length))
    autocorrelations_mean = torch.mean(autocorrelations, dim=(1, -1))  # bsz x tgt_len
    top_k_autocorrelations, top_k_delays = torch.topk(autocorrelations_mean, top_k, dim=1)

    # apply softmax on the channel dim
    top_k_autocorrelations = torch.softmax(top_k_autocorrelations, dim=-1)  # bsz x top_k

    # compute aggregation: value_states.roll(delay) * top_k_autocorrelations(delay)
    delays_agg = torch.zeros_like(value_states).float()  # bsz x time_length x channel
    for i in range(top_k):
        value_states_roll_delay = value_states.roll(shifts=-int(top_k_delays[i]), dims=1)
        top_k_at_delay = top_k_autocorrelations[:, i]
        # aggregation
        top_k_resized = top_k_at_delay.view(-1, 1, 1).repeat(num_heads, tgt_len, channel)
        delays_agg += value_states_roll_delay * top_k_resized

    attn_output = delays_agg.contiguous()
    return attn_output
```
   
We did it! The Autoformer model in [now available](https://huggingface.co/docs/transformers/main/en/model_doc/autoformer) in the 🤗 Transformers library, and simply called `AutoformerModel`.

> Our strategy with this model is to show the performance of the univariate transformer models in comparison to the DLinear model which is inherently univariate. We will also present the results from _two_ multivariate transformers models trained on the same data.

Reviewer 3: ```suggestion
Our strategy with this model is to show the performance of the univariate Transformer models in comparison to the DLinear model which is inherently univariate. We will also present the results from _two_ multivariate Transformer models trained on the same data.
```

## DLinear 

The DLinear model first calculates the moving average (or trend) of the incoming signal using a certain kernel size and decomposes the signal into the residual (the seasonality) and trend via:

```python
class SeriesDecomp(nn.Module):
    def __init__(self, kernel_size):
        super().__init__()
        self.moving_avg = MovingAvg(kernel_size, stride=1)

    def forward(self, x):
        moving_mean = self.moving_avg(x)
        res = x - moving_mean
        return res, moving_mean
```

> Then two linear layers are deployed which in the point forecasting case project the context length array to a prediction length array for the seasonal and trend part and the resulting output is added together as the final prediction. In the probabilistic setting one can project the context length arrays to a `prediction-length * hidden` dim and then the resulting outputs are reshaped to `(prediction_length, hidden)`. Then a probabilistic head maps the latent representations of size `hidden` to the parameters of some distribution.

Author: ok

For our benchmark, we will use the implementation of DLinear from GluonTS.

## Benchmarking - Transformers vs. DLinear

We want to show empirically the performance of Transformer based models, by benchmarking on the `traffic` dataset, in the univariate setting. We will keep the following hyper-parameters fixed for all the models:
* `context_length = prediction_length*2`
* `batch_size = 128`
* `num_batches_per_epoch = 100`
* `epochs = 50`
* `scaling = "std"`

The transformers models are all relatively small  with:
* `encoder_layers=2`
* `decoder_layers=2`
* `d_model=16`

Instead of showing how to train a model using `Autoformer` one can just replace the model in the previous two blog posts with the new `Autoformer` model and train it on the `traffic` dataset. In order to not repeat ourself, we  have done train the models already and pushed them to the Hugging Face Hub. In the following we will use the already saved models for just the evaluations.

## Set-up Environment

> First, let's install the necessary libraries: 🤗 Transformers, 🤗 Datasets, 🤗 Evaluate, 🤗 Accelerate and [GluonTS](https://github.com/awslabs/gluonts).

Reviewer 3: ```suggestion
First, let's install the necessary libraries: 🤗 Transformers, 🤗 Datasets, 🤗 Evaluate, 🤗 Accelerate and GluonTS.
```
Have added the link above

As we will show, GluonTS will be used for transforming the data to create features as well as for creating appropriate training, validation and test batches.

```python
%pip install -q git+https://github.com/huggingface/transformers.git datasets evaluate accelerate "gluonts[pro,torch]" ujson tqdm
```

We also quickly upload some telemetry - this tells us which examples and software versions are getting used so we know where to prioritize our maintenance efforts. We don't collect (or care about) any personally identifiable information, but if you'd prefer not to be counted, feel free to skip this step or delete this cell entirely.

```python
from transformers.utils import send_example_telemetry

send_example_telemetry("autoformer_notebook", framework="pytorch")

> ```

Author: ah yes! missed it! removing

```python
def plot_gluonts(index):
    plt.plot(d_linear_tss[index][-4 * dataset.metadata.prediction_length:].to_timestamp(), label="target")
    d_linear_forecasts[index].plot(show_label=True,  color='g')
    plt.legend()
    plt.gcf().autofmt_xdate()
    plt.show()
```

```python
plot_gluonts(4)
```

![png](https://huggingface.co/datasets/huggingface/documentation-images/resolve/main/blog/148_autoformer/output_54_0.png)
    

The `traffic` dataset has a distributional shift in the sensor patterns between weekdays and weekends. So what is going on here? Since the DLinear model has no capacity to incorporate covariates, in particular any date-time features, the context window we give it does not have enough information to figure out if the prediction is for the weekend or weekday. Thus the model will predict the more common of the patterns, namely the weekdays leading to poorer performance on weekends. Of course, by giving it a larger context window a linear model will figure out the weekly pattern but perhaps there is a monthly or quaterly pattern in the data which would require bigger and bigger contexts.

## Conclusion

> How do transformer based models comapre against the above linear baseline? The test set MASE metrics from the different models we have are below:

Reviewer 3: ```suggestion
How do Transformer based models compare against the above linear baseline? The test set MASE metrics from the different models we have are below:
```

|Dataset | 	 Transformer (uni.) |   	 Transformer (mv.)  | Informer (uni.)| Informer (mv.) | Autoformer (uni.) | DLinear |
|:--:|:--:| :--:| :--:| :--:|  :--:|  :--:| 
|`Traffic` 	| **0.876** | 1.046 | 0.924 | 1.131  | 0.910 | 0.969 |

> As one can observe, the multivariate forecasts are typically _worse_ than the univariate ones, the reason being the difficulty in estimating the cross-series correlations/relationships. The additional variance added by the estimates often harms the resulting forecasts or the model learns spurious correlations. We refer to [this paper](https://openreview.net/forum?id=GpW327gxLTF) for further reading. Multivariate models tend to work well when trained on a lot of data. However, when one compares  univariate models with multivariate models, especially on smaller open datasets, the univariate models give better metrics. By comparing the linear model with equivalent-sized univariate transformers or in fact any other neural univariate model, one will typically get better performance.

Reviewer 3: ```suggestion
As one can observe, the [vanilla Transformer](https://huggingface.co/docs/transformers/model_doc/time_series_transformer) which we introduced last year gets the best results. Secondly, multivariate forecasts are typically _worse_ than the univariate ones, the reason being the difficulty in estimating the cross-series correlations/relationships. The additional variance added by the estimates often harms the resulting forecasts or the model learns spurious correlations. We refer to [this paper](https://openreview.net/forum?id=GpW327gxLTF) for further reading. Multivariate models tend to work well when trained on a lot of data. However, when one compares univariate models with multivariate models, especially on smaller open datasets, the univariate models give better metrics. By comparing the linear model with equivalent-sized univariate transformers or in fact any other neural univariate model, one will typically get better performance.
```
